### PR TITLE
[SPARK-35538][SQL] Migrate transformAllExpressions call sites to use transformAllExpressionsWithPruning

### PR DESCRIPTION
**Pull request by sigmod**
### What changes were proposed in this pull request?

Added the following TreePattern enums:
- EXCHANGE
- IN_SUBQUERY_EXEC
- UPDATE_FIELDS

Migrated `transformAllExpressions` call sites to use `transformAllExpressionsWithPruning`

### Why are the changes needed?

Reduce the number of tree traversals and hence improve the query compilation latency.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Existing tests.
Perf diff:
Rule name | Total Time (baseline) | Total Time (experiment) | experiment/baseline
OptimizeUpdateFields | 54646396 | 27444424 | 0.5
ReplaceUpdateFieldsExpression  | 24694303 | 2087517 | 0.08

### CODE REVIEW

**Comment by sigmod** [comment]
@dbaliafroozeh @gengliangwang @hvanhovell @maryannxue  this PR is ready for review.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/subquery.scala [hunk 1]
```diff
@@ -138,11 +138,12 @@ object SubExprUtils extends PredicateHelper {
    * of the input logical plan.
    */
   def stripOuterReferences(p: LogicalPlan): LogicalPlan = {
-    p.transformAllExpressions {
+    p.transformAllExpressionsWithPruning(_.containsPattern(OUTER_REFERENCE)) {
       case OuterReference(a) => a
     }
   }
 
+
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/subquery.scala:146]: unnecessary change.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/ReuseAdaptiveSubquery.scala [hunk 2]
```diff
@@ -20,6 +20,7 @@ package org.apache.spark.sql.execution.adaptive
 import scala.collection.concurrent.TrieMap
 
 import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.catalyst.trees.TreePattern._
```
Comment (gengliangwang) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/ReuseAdaptiveSubquery.scala:23]: nit: let's just import org.apache.spark.sql.catalyst.trees.TreePattern.PLAN_EXPRESSION

**Review by gengliangwang — APPROVED** [review]
LGTM

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/subquery.scala (hunk 1 above)
Comment (sigmod) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/subquery.scala]: Done.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/ReuseAdaptiveSubquery.scala (hunk 2 above)
Comment (sigmod) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/ReuseAdaptiveSubquery.scala]: Done.

**Review by jiangxb1987 — APPROVED** [review]
LGTM

**Comment by jiangxb1987** [comment]
Thanks, merged to master!
